### ISSUE TITLE
[js] Using a StringMap in __init__

### ISSUE THREAD

**Issue by benmerckx** [HaxeFoundation/haxe#5707]
The `__init__` static of a StringMap is called after an `__init__` which uses a StringMap. Resulting in `__map_reserved` being undefined:
```
Cannot read property 'key' of undefined
```
Example (see compiled js): http://try.haxe.org/#241E8
```haxe
class Test {
    static function __init__() {
        var a = new Map();
        a.set('key', 'value');
        trace(a.get('key'));
    }
}
```

**Comment by Simn** [comment]
I wouldn't do something like that in `__init__` in the first place, Statics aren't initialized at that point, so calling the constructor of something might cause undefined behavior.

Having said that I'm not sure why the order isn't right here. Seems to be a case that we could detect. Does it also fail if you use StringMap directly instead of Map?

**Comment by benmerckx** [comment]
Well, my [use case](https://github.com/benmerckx/dotenv/blob/master/tests/RunTests.hx#L13) (get some variables from env or local .env file) is about initializing statics :)

Using StringMap directly fixes the ordering indeed.

**Comment by ncannasse** [comment]
Yes __init__ is often thought as a possibility to initialize statics, whereas it's used for low level class setup for libraries/compiler devs. I think we don't document it atm, maybe we should make that more clear in the HaxeManual?

**Comment by Simn** [comment]
Agreed, there used to be this boot magic section in the old wiki. We'll have to check how accurate that information still is.

If we document this, we should also mention an alternative. I've seen people use `static var _ = { init code }` which I think is not a bad idea.

**Comment by Simn** [comment]
Oh and the Map issue is gonna be annoying. We run abstract specialization during filters, but the type ordering is done earlier in `Typer.generate` which simply doesn't know yet that we're having a StringMap instance. Not a quick fix to make.

**Comment by benmerckx** [comment]
I have used `__init__` for that purpose more than once, thanks for clearing things up and providing an alternative.

**Comment by RealyUniqueName** [comment]
What if we move the code of `StringMap.__init__` to the constructor?
https://github.com/HaxeFoundation/haxe/blob/ce5ce1979df53486256dc872e748c8fe4dfb697b/std/js/_std/haxe/ds/StringMap.hx#L154-L156
https://github.com/HaxeFoundation/haxe/blob/ce5ce1979df53486256dc872e748c8fe4dfb697b/std/js/_std/haxe/ds/StringMap.hx#L48-L50
change to
```haxe
	public inline function new() : Void {
		if(!js.Syntax.code("$global.__map_reserved")) 
			js.Syntax.code("$global.__map_reserved = {}");
		h = {};
	}
```
Doesn't look like it will impact performance.

**Comment by nadako** [comment]
Maybe we should have some option to inject custom init js code into the beginning of the generated JS, rather than the very end....

**Comment by RealyUniqueName** [comment]
That wouldn't solve the issue. It would still rely on the order of execution of such code.

**Comment by ncannasse** [comment]
I would vote for a wontfix here for reasons explained before
